lib/src/async_import_cache.dart

```diff
@@ -154,64 +154,79 @@ final class AsyncImportCache {
     }
 
     if (baseImporter != null && url.scheme == '') {
-      var relativeResult = await putIfAbsentAsync(
-          _relativeCanonicalizeCache,
-          (
-            url,
-            forImport: forImport,
-            baseImporter: baseImporter,
-            baseUrl: baseUrl
-          ),
-          () => _canonicalize(baseImporter, baseUrl?.resolveUri(url) ?? url,
-              baseUrl, forImport));
+      var relativeResult = await putIfAbsentAsync(_relativeCanonicalizeCache, (
+        url,
+        forImport: forImport,
+        baseImporter: baseImporter,
+        baseUrl: baseUrl
+      ), () async {
+        var (result, cacheable) = await _canonicalize(
+            baseImporter, baseUrl?.resolveUri(url) ?? url, baseUrl, forImport);
+        assert(
+            cacheable,
+            "Relative loads should always be cacheable because they never "
+            "provide access to the containing URL.");
+        return result;
+      });
       if (relativeResult != null) return relativeResult;
     }
 
-    return await putIfAbsentAsync(
-        _canonicalizeCache, (url, forImport: forImport), () async {
-      for (var importer in _importers) {
-        if (await _canonicalize(importer, url, baseUrl, forImport)
-            case var result?) {
+    var key = (url, forImport: forImport);
+    if (_canonicalizeCache.containsKey(key)) return _canonicalizeCache[key];
```

Comment:
I can think of an edge case with two importers `[importerA, importerB]`:

1. Resolve an import `x`, with containing url `u:y`. `importerA` accessed `containingUrl`, but did not resolve it, then `importerB` resolved it without using `containingUrl` thus get cached.
2. Resolve an import `x`, with containing url `file://a.scss`. `importerA` would have been able to resolve it because now this has a different `containingUrl` that `importerA` can handle, however, because of the previous cache from `importerB`, `importerA` would not even get attempted.

I think this would be unexpected, because `importerA` is before `importerB`.

Reply:
In this case, `_canonicalize()` would return `(null, false)` for because `containingUrl` is passed, so `cacheable` will be set to `false` on line 188 and `importerB`'s result won't be cached even if it would be cacheable in isolation.

Reply:
The expectation here is that the result with cache should be exactly the same as if we completely remove the cache capability, thus the order of importers would matter. Implement the order meaning we need cache per importer here, which would can have performance impact if a non-cacheable importer has high priority and low hit rate...

However, for "correctness", it seems to be unavoidable.

Reply:
Sorry, I forgot this is the first iteration that cacheable is based on whether it's passed or not, instead of whether it's used or not.

Reply:
The logic should work the same if it just checks access—we'd still return `(null, true)` in the case where `importerA` accessed `containingUrl` but couldn't resolve the import.

Reply:
I think this works correctly.

If I read it correctly, any non-cacheable importer will make all importer after it non-cacheable for the current import. And therefore if a FilesystemImporter is low priority it would not be cached and cost can be high? If that’s the case it might still be worth to have per importer cache key instead of single cache key for all importers.

Reply:
If we have `[nonCacheableImporter, loadPathImporter]`, with the implementation in this PR the `loadPathImporter` would not be cached, and repeatedly loading the same file from load path can get slower due to repeated I/O calls.

Thus I think it's worth to make the trade-off to use a little bit more memory to cache per importer (and even cache failed canonicalization if cacheable), and a bit more CPU to re-check cache per importer, like the pseudo code below:  

``` dart
for (var importer in importers) {
    var key = (url, forImport: forImport, importer: importer);
    if (_canonicalizeCache.containsKey(key)) {
      var cached = _canonicalizeCache[key];
      if (cached != null) return cached;
    } else {
      var (result, cacheable) = await _canonicalize(importer, url, baseUrl, forImport);
      if (cacheable) {
        _canonicalizeCache[key] = result;
      }
      if (result != null) return result;
    }
}
return null;
```

The cost of CPU/memory overhead would likely be lower than the I/O overhead.

Reply:
I don't think it's crazy to have a per-importer canonicalize cache in addition to the current whole-load-path cache, although I think we might want to be a bit more sophisticated about only filling it if we run into an uncacheable load in practice. Either way, let's save that for a follow-up after we fix the initial bug.